File: infrastructures/ddd/python/src/infrastructure/queues/infra/i_event_subscriber.py
"""
Interface for subscribing to domain events from a message queue.
"""

from abc import ABC, abstractmethod
from typing import Awaitable, Callable, TypeVar


TEvent = TypeVar('TEvent')
EventHandler = Callable[[TEvent], Awaitable[None]]


class IEventSubscriber(ABC):
    """
    Interface for subscribing to domain events.
    
    Subscribers:
    - Listen to specific topics
    - Handle events asynchronously
    - Can trigger side effects in their subdomain
    """
    
    @abstractmethod
    async def subscribe_async(
        self,
        p_topic: str,
        p_handler: EventHandler[TEvent]
    ) -> None:
        """
        Subscribes to events on the specified topic.
        
        Args:
            p_topic: The topic to subscribe to
            p_handler: The async handler function for events
        """
        pass

File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/infrastructure/queues/infra/IEventSubscriber.java
package com.oddly.ddd.infrastructure.queues.infra;

import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

/**
 * Interface for subscribing to domain events from a message queue.
 * Event subscribers:
 * - Live in /infrastructure/queues/subscribers/
 * - Handle events from other subdomains
 * - Process events asynchronously
 * 
 * @param <TEvent> The event type to subscribe to
 */
public interface IEventSubscriber<TEvent> {
    /**
     * Subscribes to events on the specified topic.
     * 
     * @param p_topic The topic to subscribe to
     * @param p_handler The handler function to process events
     */
    CompletableFuture<Void> subscribeAsync(
        String p_topic,
        Function<TEvent, CompletableFuture<Void>> p_handler
    );

    /**
     * Handles an event.
     * Implement this method to process the event.
     * 
     * @param p_event The event to handle
     */
    CompletableFuture<Void> handleAsync(TEvent p_event);
}
